File: querier/internal/paths/paths.go
package paths

import (
	"fmt"
	"path/filepath"
	"runtime"
)

type DataPaths struct {
	Home     string
	Manifest string
	Shards   string
	Sessions string
}

func Resolve(explicitHome string) (DataPaths, error) {
	home, err := userHome()
	if err != nil {
		return DataPaths{}, err
	}
	return ResolveFor(explicitHome, environment(), runtime.GOOS, home)
}

func ResolveFor(explicitHome string, env map[string]string, goos, userHome string) (DataPaths, error) {
	var home string
	switch {
	case explicitHome != "":
		home = explicitHome
	case env["RESEARCH4JAR_HOME"] != "":
		home = env["RESEARCH4JAR_HOME"]
	case goos == "darwin":
		home = filepath.Join(userHome, "Library", "Application Support", "research4jar")
	case goos == "windows":
		if env["LOCALAPPDATA"] == "" {
			return DataPaths{}, fmt.Errorf(
				"LOCALAPPDATA is required on Windows when RESEARCH4JAR_HOME is unset",
			)
		}
		home = filepath.Join(env["LOCALAPPDATA"], "research4jar")
	default:
		if env["XDG_DATA_HOME"] != "" {
			home = filepath.Join(env["XDG_DATA_HOME"], "research4jar")
		} else {
			home = filepath.Join(userHome, ".local", "share", "research4jar")
		}
	}

	absolute, err := filepath.Abs(home)
	if err != nil {
		return DataPaths{}, fmt.Errorf("resolve Research4Jar home: %w", err)
	}
	return DataPaths{
		Home:     absolute,
		Manifest: filepath.Join(absolute, "manifest.db"),
		Shards:   filepath.Join(absolute, "shards"),
		Sessions: filepath.Join(absolute, "sessions"),
	}, nil
}

File: core/src/main/kotlin/dev/research4jar/indexer/Paths.kt
package dev.research4jar.indexer

import dev.research4jar.runtime.WorkingDirectoryContext
import java.nio.file.Path
import java.nio.file.Paths

data class DataPaths(
    val home: Path,
    val manifest: Path = home.resolve("manifest.db"),
    val shards: Path = home.resolve("shards"),
    val sessions: Path = home.resolve("sessions"),
)

object Research4JarPaths {
    fun resolve(
        explicitHome: String? = null,
        environment: Map<String, String> = System.getenv(),
        osName: String = System.getProperty("os.name"),
        userHome: String = System.getProperty("user.home"),
    ): DataPaths {
        val home = when {
            !explicitHome.isNullOrBlank() -> Paths.get(explicitHome)
            !environment["RESEARCH4JAR_HOME"].isNullOrBlank() ->
                Paths.get(environment.getValue("RESEARCH4JAR_HOME"))

            osName.lowercase().let { it.contains("mac") || it.contains("darwin") } ->
                Paths.get(userHome, "Library", "Application Support", "research4jar")

            osName.lowercase().startsWith("windows") -> {
                val localAppData = environment["LOCALAPPDATA"]
                    ?: error("LOCALAPPDATA is required on Windows when RESEARCH4JAR_HOME is unset")
                Paths.get(localAppData, "research4jar")
            }

            else -> {
                val xdg = environment["XDG_DATA_HOME"]
                if (xdg.isNullOrBlank()) {
                    Paths.get(userHome, ".local", "share", "research4jar")
                } else {
                    Paths.get(xdg, "research4jar")
                }
            }
        }
        return DataPaths(WorkingDirectoryContext.resolve(home))
    }
}
